401 fix

yigith opened the pull request:
#401 

/Basket/Checkout has been converted to a read-only version of /Basket/Index as discussed in #401 .

Both /Basket/Index and /Basket/Checkout needs some handling about their basket items layout. When the browser is resized, its layout crashes. I preserved the existing structure and layout.

Next week, I can also work on the Layout if it's OK for you.

Note: I've tested the basket operations against the cases mentioned in #401 . It looks working well.

efleming18 commented on src/ApplicationCore/Services/BasketService.cs:
```diff
@@ -59,11 +59,22 @@ public async Task TransferBasketAsync(string anonymousId, string userName)
-            if (basket == null) return;
-            basket.SetNewBuyerId(userName);
-            await _basketRepository.UpdateAsync(basket);
+            var anonymousBasketSpec = new BasketWithItemsSpecification(anonymousId);
```
Can we add some tests here? There should already be a test file for `TransferBasketAsync`, it just needs updated to account for the new logic.

yigith replied:
I plan to write some tests related to following:

- removes anonymous basket after transfer
- preserves existing basket items in the user basket while transferring anonymous basket items 
- creates a user basket before transferring if not exists

efleming18 replied:
Those sound great! You could also write a few unit tests that mock out the `_basketRepository` and ensure all the various methods are invoked the number of times you expect for the different scenarios.

yigith replied:
```
        [Fact]
        public async Task InvokesBasketRepositoryFirstOrDefaultAsyncOnceIfAnonymousBasketNotExists()
        {
            Basket basket = null;
            _mockBasketRepo.Setup(x => x.FirstOrDefaultAsync(It.IsAny<BasketWithItemsSpecification>())).ReturnsAsync(basket);
            var basketService = new BasketService(_mockBasketRepo.Object, null);
            await basketService.TransferBasketAsync("AnIdThatDoesNotExist", "testuser@microsoft.com");
            _mockBasketRepo.Verify(x => x.FirstOrDefaultAsync(It.IsAny<BasketWithItemsSpecification>()), Times.Once);
        }
```

I might use your opinion on this. Here, I am trying to confirm that it doesn't look for or create a user basket if anonymous basket to transfer from doesn't exist.

efleming18 replied:
That is looking good! You could update the verify to check that the specific value was used when you tried to invoke it. So instead of passing in `It.IsAny<BasketWithItemsSpecification>()` you could create the two different specifications and pass them in. That way you can actually have multiple verifies on the `FirstOrDefaultAsync` call.

One that confirms that `FirstOrDefaultAsync` was invoked with the `anonymousBasketSpec`, and another that confirms `FirstOrDefaultAsync` was **never** invoked for the `userBasketSpec`.

yigith replied:
```csharp
private readonly string _nonexistentAnonymousBasketBuyerId = "nonexistent-anonymous-basket-buyer-id";
private readonly string _existentUserBasketBuyerId = "testuser@microsoft.com";
[Fact]
public async Task InvokesBasketRepositoryFirstOrDefaultAsyncOnceIfAnonymousBasketNotExists()
{
    var nullBasket = null as Basket;
    var userBasket = new Basket(_existentUserBasketBuyerId);
    var anonymousBasketSpec = new BasketWithItemsSpecification(_nonexistentAnonymousBasketBuyerId);
    var userBasketSpec = new BasketWithItemsSpecification(_existentUserBasketBuyerId);
    _mockBasketRepo.Setup(x => x.FirstOrDefaultAsync(anonymousBasketSpec)).ReturnsAsync(nullBasket);
    _mockBasketRepo.Setup(x => x.FirstOrDefaultAsync(userBasketSpec)).ReturnsAsync(userBasket);
    var basketService = new BasketService(_mockBasketRepo.Object, null);
    await basketService.TransferBasketAsync("AnIdThatDoesNotExist", "testuser@microsoft.com");
    _mockBasketRepo.Verify(x => x.FirstOrDefaultAsync(anonymousBasketSpec), Times.Once);
    _mockBasketRepo.Verify(x => x.FirstOrDefaultAsync(userBasketSpec), Times.Never);
}
```

I've just prepared this faulty code to be able to discuss on it. Something like this doesn't work because of 2 main problems:
1. Last setup overrides the first setup as explained on the explanation of the method: If more than one setup is specified for the same method or property, the latest one wins and is the one that will be executed.
2. instead of x => Since BasketWithItemsSpecification is a reference type, it will not match with the specification object created in the TransferBasketAsync service method. If an equality checker existed for the type ISpecification, I thought I could update ReturnsAsync and Verify part so that it will return the right basket for each specification.

efleming18 replied:
Ah, yes - I see what you mean. For now, we don't have to do the verify of the specific Specification but let's still do the verification of the number of times the method was called. 

For issue #1 you described above, you can actually mock out a "sequence" like this:

```
            _mockBasketRepo.SetupSequence(x => x.FirstOrDefaultAsync(It.IsAny<BasketWithItemsSpecification>()))
                .ReturnsAsync(anonymousBasket)
                .ReturnsAsync(userBasket);
```

The above snippet will return an `anonymousBasket` on the first time the `FirstOrDefaultAsync` is invoked, and will return the different `userBasket` the second time it's invoked. Hopefully that helps!

yigith replied:
Thank you @efleming18, your suggestion helped. I will proceed accordingly.

yigith replied:
https://github.com/dotnet-architecture/eShopOnWeb/pull/408/commits/9fdb60db714df7b15cde768cb7fc5d659043839b

efleming18 commented on src/Web/Areas/Identity/Pages/Account/Login.cshtml.cs:
```diff
@@ -99,5 +103,15 @@ public async Task<IActionResult> OnPostAsync(string returnUrl = null)
             return Page();
         }
+
+        private async Task TransferAnonymousBasketToUserAsync(string userName)
```
Would you be able to write some tests around this as well? This should also probably be pulled out in to the `BasketService`

yigith replied:
This method is just calling the TransferBasketAsync method of the basket service in case a cookie with anonymous buyer id exists. I created this private method so that its content wouldn't make the OnPostAsync method fatter.

```csharp
        private async Task TransferAnonymousBasketToUserAsync(string userName)
        {
            if (Request.Cookies.ContainsKey(Constants.BASKET_COOKIENAME))
            {
                var anonymousId = Request.Cookies[Constants.BASKET_COOKIENAME];
                await _basketService.TransferBasketAsync(anonymousId, userName);
                Response.Cookies.Delete(Constants.BASKET_COOKIENAME);
            }
        }
```
This method is private and contains web elements. 

> This should also probably be pulled out in to the BasketService
By this, did you mean that this should be in the ApplicationCore/BasketService instead of Login.cshtml.cs?

efleming18 replied:
I see now that it's dealing mostly with Web elements, and calling the existing method on the `BasketService`. I'm ok with how you've done this. No changes required for this.

efleming18 commented:
@yigith Could you also rebase master in to your branch here? There were a few unit tests that were failing (unrelated to your changes) that I fixed in a recent PR. Before adding the tests I was asking about above, if you just make sure this branch is up to date, you should be good to move forward.

yigith commented:
> @yigith Could you also rebase master in to your branch here? There were a few unit tests that were failing (unrelated to your changes) that I fixed in a recent PR. Before adding the tests I was asking about above, if you just make sure this branch is up to date, you should be good to move forward.

@efleming18 I rebased my branch onto the master and now working on the unit tests. I have a few questions about them. I will ask them above under change requests.

efleming18 commented:
This is really great stuff here @yigith! Thank you for your continued contributions!

yigith commented:
> This is really great stuff here @yigith! Thank you for your continued contributions!

You are welcome @efleming18. I am glad to contribute. Thank you for your kind words and technical support during this process.
